### PR TITLE
`azurerm_batch_pool` - Added support for `identity_reference` into `container_registries`

### PR DESCRIPTION
**Pull request by dkuzmenok**
Added support for using `identity_reference` inside of `container_registries` of `azurerm_batch_pool`.

API link: https://docs.microsoft.com/en-us/rest/api/batchservice/pool/add#containerregistry

Test
----------
```
$ go test -v -timeout 3000s -run ^TestAccBatchPool_container github.com/hashicorp/terraform-provider-azurerm/internal/services/batch/
=== RUN   TestAccBatchPool_containerWithUser
=== PAUSE TestAccBatchPool_containerWithUser
=== RUN   TestAccBatchPool_containerWithUAMI
=== PAUSE TestAccBatchPool_containerWithUAMI
=== CONT  TestAccBatchPool_containerWithUser
=== CONT  TestAccBatchPool_containerWithUAMI
--- PASS: TestAccBatchPool_containerWithUser (587.65s)
--- PASS: TestAccBatchPool_containerWithUAMI (589.87s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/batch 593.975s
```

### CODE REVIEW

**Comment by dkuzmenok** [comment]
Not sure if there is something to change after those failed tests. 
It complains about the use of an internal package, used in other places in the same way...

File: internal/services/batch/batch_pool.go [hunk 1]
```diff
@@ -363,11 +361,27 @@ func expandBatchPoolContainerRegistry(ref map[string]interface{}) (*batch.Contai
...
-	containerRegistry := batch.ContainerRegistry{
-		RegistryServer: utils.String(ref["registry_server"].(string)),
-		UserName:       utils.String(ref["user_name"].(string)),
-		Password:       utils.String(ref["password"].(string)),
+	containerRegistry := batch.ContainerRegistry{}
+
+	if ref["registry_server"] != nil && ref["registry_server"] != "" {
+		refRegistryServer := ref["registry_server"].(string)
+		containerRegistry.RegistryServer = &refRegistryServer
+	}
```
Comment (katbyte) [inline internal/services/batch/batch_pool.go:369]: minor: could we change these to be something more like this
```suggestion
	if v := ref["registry_server"]; v != nil && v != "" {
		containerRegistry.RegistryServer = pointer.FromString(v.(string))
	}
```

File: internal/services/batch/batch_pool_resource.go [hunk 2]
```diff
@@ -159,15 +160,22 @@ func resourceBatchPool() *pluginsdk.Resource {
 										ForceNew:     true,
 										ValidateFunc: validation.StringIsNotEmpty,
 									},
+									"identity_reference": {
```
Comment (katbyte) [inline internal/services/batch/batch_pool_resource.go:163]: if this is always the user assigned identity could we make it
```suggestion
									"user_assigned_identity_id": {
```

File: internal/services/batch/batch_pool_resource.go [hunk 3]
```diff
@@ -159,15 +160,22 @@ func resourceBatchPool() *pluginsdk.Resource {
...
+										Optional:     true,
+										ForceNew:     true,
+										ValidateFunc: msivalidate.UserAssignedIdentityID,
+										Description:  "The User Assigned Identity to use for Container Registry access.",
+									},
 									"user_name": {
 										Type:         pluginsdk.TypeString,
-										Required:     true,
+										Optional:     true,
 										ForceNew:     true,
```
Comment (katbyte) [inline internal/services/batch/batch_pool_resource.go:173]: can you specify both user + password and the identity? if not could we add some validation via ExactlyOneOf
```suggestion
										ForceNew:     true,
										ExactlyOneOf
```

File: internal/services/batch/batch_pool_resource.go [hunk 4]
```diff
@@ -159,15 +160,22 @@ func resourceBatchPool() *pluginsdk.Resource {
...
 										ValidateFunc: validation.StringIsNotEmpty,
 									},
 									"password": {
 										Type:         pluginsdk.TypeString,
-										Required:     true,
+										Optional:     true,
 										ForceNew:     true,
```
Comment (katbyte) [inline internal/services/batch/batch_pool_resource.go:179]: and I presume this is required when user is so
```suggestion
										ForceNew:     true,
										RequiredWith: []{"user"},
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @dkuzmenok - overall looks good and I've left some suggestions inline to address before getting this merged

File: internal/services/batch/batch_pool_resource.go (hunk 4 above)
Comment (dkuzmenok) [reply internal/services/batch/batch_pool_resource.go:179]: Here is the existing configuration, exactly with the idea as mine: https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/containers/container_registry_task_resource.go#L532
(Have not seen it before).

**Comment by dkuzmenok** [comment]
@katbyte Thanks for your review. I have pushed changes to resolve your comments. 

The only one left is that question with `ExactlyOneOf`, but, please, read my comments on that.

**Review by katbyte — APPROVED** [review]
Thanks @dkuzmenok - LGTM

**Pull request merged**